dev/tools/controllerbuilder/pkg/updatetypes/main.go

```diff
@@ -0,0 +1,235 @@
+	flag.StringVar(&opts.apiDirectory, "api-dir", "../../../../../apis/", "Base directory for APIs")
+	flag.StringVar(&opts.ignoredFields, "ignored-fields", opts.ignoredFields, "Comma-separated list of fields to ignore")
+	flag.Parse()
+	return opts
+}
+
+func NewTypeUpdater(opts options, protoMsgsNotMappedToGoStruct []string) *TypeUpdater {
+	return &TypeUpdater{
+		opts:                         opts,
+		protoMsgsNotMappedToGoStruct: protoMsgsNotMappedToGoStruct,
+	}
+}
+
+func (u *TypeUpdater) Run() error {
+	parentMsg, newField, err := findNewField(u.opts.protoSourcePath, u.opts.parentMessageFullName, u.opts.newField)
+	if err != nil {
+		return err
+	}
+
+	msgs, err := findDependentMsgs(newField, strings.Split(u.opts.ignoredFields, ","))
+	if err != nil {
+		return err
+	}
+
+	removeNotMappedToGoStruct(msgs)
+
+	if err := removeAlreadyGenerated(u.opts.goPackagePath, u.opts.apiDirectory, msgs); err != nil {
+		return err
+	}
+
+	// TODO: find the proper file to write/update. Currently print to a tmp file.
+	return writeToLocalFile(newField, parentMsg, msgs)
+}
+
+// findNewField locates the parent message and the new field in the proto file
+func findNewField(pbSourcePath, parentMsgFullName, newFieldName string) (protoreflect.MessageDescriptor, protoreflect.FieldDescriptor, error) {
+	fileData, err := os.ReadFile(pbSourcePath)
+	if err != nil {
+		return nil, nil, fmt.Errorf("reading %q: %w", pbSourcePath, err)
+	}
+
+	fds := &descriptorpb.FileDescriptorSet{}
+	if err := proto.Unmarshal(fileData, fds); err != nil {
+		return nil, nil, fmt.Errorf("unmarshalling %q: %w", pbSourcePath, err)
+	}
+
+	files, err := protodesc.NewFiles(fds)
+	if err != nil {
+		return nil, nil, fmt.Errorf("building file description: %w", err)
+	}
+
+	// Find the parent message
+	messageDesc, err := files.FindDescriptorByName(protoreflect.FullName(parentMsgFullName))
+	if err != nil {
+		return nil, nil, err
+	}
+	msgDesc, ok := messageDesc.(protoreflect.MessageDescriptor)
+	if !ok {
+		return nil, nil, fmt.Errorf("unexpected descriptor type: %T", msgDesc)
+	}
+
+	// Find the new field in parent message
+	fieldDesc := msgDesc.Fields().ByName(protoreflect.Name(newFieldName))
+	if fieldDesc == nil {
+		return nil, nil, fmt.Errorf("field not found in message")
+	}
+
+	return msgDesc, fieldDesc, nil
+}
+
+// findDependentMsgs finds all dependent proto messages for the given field, ignoring specified fields
+func findDependentMsgs(field protoreflect.FieldDescriptor, ignoredProtoFields []string) (map[string]protoreflect.MessageDescriptor, error) {
+	deps := make(map[string]protoreflect.MessageDescriptor)
+	analyzeDependencies(field, deps, ignoredProtoFields)
+
+	// fmt.Println("\nDependencies:")
+	// for _, dep := range deps {
+	// 	fmt.Printf("%s\n", dep.FullName())
+	// }
+	return deps, nil
+}
+
+// analyzeDependencies recursively explores the dependent proto messages of the given field.
```

Comment:
Oooh this could be really useful.  Technically we only want to generate "reachable" messages from the "main resource proto".  Today, we just generate them all, which is overkill, and sometimes we have to fix bugs in messages we don't use anyway.

One for the future though, rather than this PR :-)